All parsing of arguments now split out.

stevepentland opened the pull request:
Also, splits out building HyperfineOptions, Commands vec, and ExportManager.

This now means we can test the following functionality:

- Command line to expected matches
- Matches to ```HyperfineOptions```
- Matches to ```Vec<Command>```
- Matches to ```ExportManager```

So we can ensure that changes do not unexpectedly alter how we handle these elements. I will be opening future PRs to work on these tests, split by the domains listed above.

Also added quick test to ensure the default state of ```HyperfineOptions``` does not change unexpectedly.

sharkdp commented on src/hyperfine/types.rs:
```diff
@@ -157,3 +157,66 @@ impl BenchmarkResult {
+
+/// Ensure that the default options are not changed unintentionally
+#[test]
+fn test_ensure_default_options_values() {
```
To be honest, I'm not a big fan of this unit test. It is basically just re-iterating the default values from `types.rs` which means that we would have to change default values in two places. There is hardly any functionality that is tested here.

stevepentland replied:
If we were deriving Default, I'd agree. But that's the entire point, to prevent unintended changes to a known good application state. We know and have established what the default representation of the options should be. Preventing a regression, that could be hard to pinpoint, from being introduced is worth having to update the test when an explicit change is desired no?

sharkdp commented on src/main.rs:
```diff
@@ -269,20 +191,51 @@ fn main() {
+}
+
+#[cfg(test)]
+fn build_commands_over_range(
```
It somehow feels wrong that we have to basically duplicate the functionality in order to test it. In my opinion, unit tests should be self-contained with hard-coded expected results.

stevepentland replied:
This was just done to avoid writing the same logic over and over again. We could just build the expected commands for each test (there should be a few possible ones for this particular function), which is what I originally did, but that becomes rather verbose.

sharkdp commented:
Thank you. The restructuring looks good, but to be honest, I'm not sure about the unit tests (see inline comments).

I recently learned about [assert_cli!](https://github.com/assert-rs/assert_cli). Maybe that would be something that we could use?

stevepentland commented:
> I recently learned about assert_cli!. Maybe that would be something that we could use?

For sure, but this stands as more of integration testing. It will be valuable, but we can't really embark on ensuring all the component parts of the application are properly wired up until we can assert each of those component parts is operating correctly in the first place.

stevepentland commented:
I have left in the refactor and switched `HyperfineOptions` to use only const values, which we can test against with no logic duplication. I removed the other experimental testing, but would still like to target the individual functions responsible for the important logic. Your suggestion of including `assert_cli!` is also something I think should be pursued also once individual unit testing has been expanded.

sharkdp commented:
The refactoring looks really great and I would definitely like to merge this. 

However, I have to say that I am still not in favor of the changes to `HyperfineOptions` (replacing the values by `const`ants) and the `ensure_default_options_values` test. Instead of just putting the name and default-value of a particular option in just one place (like `min_time_sec: 3.0`), I would now have to create a new constant, duplicate the names (`min_time_sec` and `DEFAULT_MIN_TIME_SEC`), look for the values at another place in the file, and write a new line in the `ensure_default_options_values` test. This looks like a lot of overhead to me for no apparent gain. How could the `default` option possibly return something else than the values that I already specify inside this function?

stevepentland commented:
> How could the default option possibly return something else than the values that I already specify inside this function?

It isn't that is could, but that an inadvertent change could potentially result in an unforeseen result. If someone edits the file and, say, changes the default failure option or default output type to something else and it is not noticed this would drastically change the way the application works. Now, to catch something like this in a unit test is a fast test that can be exactly targeted at the area where modification would occur. To catch this in integration tests requires a higher level of exclusive targeting that can be hard to execute and trust (can we guarantee that there is a command that will fail, every time, on any possible system on which it is run). It may require two more steps for new values (add const, expect const in test) but it provides a lock-in of what is expected from a 'default' options value and ensures that a change be a little more explicit by actually changing the constant value.

Otherwise, things like this rely on perfect code reviews and highly targeted integration tests. If default were derived, than I would say it should not be under test, but with a human implementation, there is always a chance a typo, local change, or other mistake could creep in. It isn't that the function could return different than it does, it is that it can return exactly what is in there, regardless of what it is or whether it was what was originally intended.

sharkdp commented:
> It isn't that is could, but that an inadvertent change could potentially result in an unforeseen result. If someone edits the file and, say, changes the default failure option or default output type to something else and it is not noticed this would drastically change the way the application works.

My point is: right now, there is only *a single source of truth* for these default values (inside the `default` function - which seems pretty descriptive to me). If someone changes a value there, I sure hope that this is intentional.
